micro-frontends-with-angular-and-stenciljs.md

```diff
@@ -0,0 +1,150 @@
+
+```bash
+npm install @nxext/stencil --save-dev
+nx g @nxext/stencil:library my-lib
+```
+
+The Nnext plugin provides a generator that allows to create our first component:
+
+```bash
+nx g @nxext/stencil:component my-header --project my-lib
+```
+
+This outputs a component skeleton like this:
+
+```ts
+@Component({
+  tag: 'my-header',
+  styleUrl: 'my-header.scss',
+  shadow: true,
+})
+export class MyHeader {
+  @Prop() first: string;
+  @Prop() middle: string;
+  @Prop() last: string;
+
+  private getText(): string {
+    return (this.first || '') + (this.middle ? ` ${this.middle}` : '') + (this.last ? ` ${this.last}` : '');
+  }
+
+  render() {
+    return <div>Hello, World! I'm {this.getText()}</div>;
+  }
+}
+```
+
+I will not be going to write a StencilJS tutorial at this place. But one thing I need to mention. Stencil provides several [output targets](https://stenciljs.com/docs/output-targets). We actually want to have a single file bundle which would be the `dist-custom-elements-bundle` out target. Unfortunately, this has been deprecated. We use its successor `dist-custom-elements`. It does basically the same thing but produces a js file per component. This is the better way because it is better to load several smaller bundles than a large one. It will also be better for leveraging `http2`. We would prefer a single file because it means less file handling in our frontend's, but we can work with that.
+
+```json
+outputTargets: [
+  {
+    type: 'dist-custom-elements',
+  },
+];
+```
+
+Finally, we are deploying these output files in an Azure Blob Storage. But any static hosting would do the job.
+
+It sounds that easy, but nobody on our team had a real experience in StencilJS and especially in lazy load them in an Angular application. One problem worth mentioning is the handling of assets. If the component uses assets that get deployed, you need to make sure to load them with an absolute path. **Relative paths get resolved to the URL where the application is deployed**. We solved the problem by compiling everything (e.g. svg icons and json translation files) into the bundle. This way, we don't need to load them during the runtime.
+
+## Consuming the WebComponents dynamically during run time
+
+After creating and deploying our WebComponents, they need to be integrated into the frontend of our applications. A simple script element loading js bundle will do the job. Then we just can use the custom element:
+
+```html
+<script type="module" src='https://cdn.jsdelivr.net/npm/my-name@0.0.1/dist/myname.js'></script>
+<my-header></my-header>
+```
+
+We found a nice library, [ANGULAR EXTENSIONS ELEMENTS](https://angular-extensions.github.io/elements/#/home), which makes it even easier to lazy load elements in Angular. It also supports displaying special components for the loading and error cases.
```

Comment:
"ANGULAR EXTENSIONS ELEMENTS" --> I think the all uppercasing is just because of the Material UI design using this in their typography, and not meant to be the casing of the library name. I would case it just with Pascal case to make it not look too weird.

Reply:
I usually try to take over the typing from the original content, but you might be right, that's caused by Angular Material.